content/blog/2019-11-02-custom-alloc/index.md

```diff
@@ -0,0 +1,206 @@
+ 
+ * Reduce the allocation time, by using only a single call to malloc, and getting enough space for them all at once (even though you might not know the exact number, or use them all at the same time, and you can't create them all yet). 
+ * Save time with free calls: you can free them all at once at a the end, rather than freeing each object individually (which could still happen regularly) during phase one of the program when all of its references go away.
+[Reconsidering Custom Memory Allocation][paper]
+Allocating memory in this way would in fact be creating a _region_ allocator: you allocate one large chunk of memory at first, and then increment a pointer into it; the entire region must then be freed all at one time. 
+
+
+#### Saving Space with Custom Allocation
+
+The example above actually uses more memory than a heap. That is more common, but they can also be used to reduce the memory footprint, primarily by preventing fragmentation. For instance, if memory fragmentation is a huge issue, but you know rougly what fraction of your memory is used by objects of different sizes, you can partition it into pieces, and deal with each size class separately, which offers much better guarantees about the worst and average case fragemtation. This is commonly done in practice; [Mesh][], for instance, requires that all pages in question are of the same size class. [The paper we analyze][paper] refers to allocators liek these as "per-class custom allocators".
+
+For all of these reasons, in 2002, it was common practice and widely condisered a good idea to write a custom memory allocation for your program to improve performance. 
+
+
+## The Drawbacks to Custom Allocation
+
+Of course, even without any experiments, it is easy to see how doing this could be a mistake: the effectiveness of any of these relies on certain assumptions about the programs, and so any modifications need to be done while keeping the custom allocator in mind. Moreover, non-standard allocation schemes make it much more difficult to use existing tools to analyze your code. In more detail:
+
+* Accidentally calling the standard `free` on a custom-allocated object could corrupt the heap and lead to errors
+* Custom memory allocation makes it impossible to use leak detection tools
+* It also prevents you from using a different custom allocator to do something else that's more valuable in the future
+* It keeps you from using garbage collection
+
+
+
+# The Paper: Reconsidering Custom Memory Management
+
+The paper at hand, [Reconsidering Custom Memory Allocation][paper], is packaged and sold as the thesis that custom memory allocators in practice do not out perform general ones. 
```

Comment:
Here and also above, please use "quotation marks" around the title to make it clearer that it's a title.